op.c

```diff
@@ -16566,71 +16636,70 @@ Perl_subsignature_append_positional(pTHX_ PADOFFSET padix, OPCODE defmode, OP *d
     yy_parser_signature *signature = PL_parser->signature;
     assert(signature);
 
-    if(signature->slurpy)
+    if(signature->slurpy) {
         yyerror("Slurpy parameter not last");
+        return;
+    }
 
-    UV argix = signature->elems;
+    if(!defexpr && signature->opt_params) {
+        yyerror("Mandatory parameter follows optional parameter");
+        return;
+    }
+
+    UV argix = signature->next_argix;
+    signature->next_argix++;
+
+    if(!padix && !defexpr)
+        /* This param has no var and no defaulting expression. There's 
+         * nothing else for us to do here.
+         */
+        return;
+
+    struct yy_parser_signature_param *param = subsignature_push_param();
+    param->argix = argix;
 
-    OP *varop = NULL;
     if(padix) {
         assert(PadnamePV(PadnamelistARRAY(PL_comppad_name)[padix])[0] == '$');
 
-        varop = newUNOP_AUX(OP_ARGELEM, 0, NULL, NULL);
-        varop->op_private |= OPpARGELEM_SV;
-        varop->op_targ = padix;
-        cUNOP_AUXx(varop)->op_aux = INT2PTR(UNOP_AUX_item *, argix);
-    }
+        param->padix = padix;
 
-    signature->elems++;
+        intro_my(); /* introduce the new pad variable now */
+    }
+    else
+        param->padix = 0;
 
     if(defexpr) {
-        signature->optelems++;
+        signature->opt_params++;
 
         if(defexpr->op_type == OP_NULL && !(defexpr->op_flags & OPf_KIDS))
         {
             /* caller passed in newOP(OP_NULL, 0), so we should not leak it */
             op_free(defexpr);
 
+            defexpr = newOP(OP_STUB, OPf_WANT_SCALAR);
+
             /* handle '$=' special case */
-            if(varop)
+            if(padix)
                 yyerror("Optional parameter lacks default expression");
```

Comment:
was `struct yy_parser_signature_param *param = subsignature_push_param();` leaked if this error branch was entered?